assets/src/edit-story/components/postLock/postLock.js

```diff
@@ -0,0 +1,142 @@
+import { useStory } from '../../app/story';
+import { useConfig } from '../../app/config';
+import PostLockFirstDialog from './postLockFirstDialog';
+import PostLockDialog from './postLockDialog';
+
+function PostLock() {
+  const {
+    actions: { getStoryLockById, setStoryLockById, deleteStoryLockById },
+  } = useAPI();
+  const {
+    storyId,
+    userId,
+    dashboardLink,
+    postLock: { interval: postLockInterval, showLockedDialog },
+  } = useConfig();
+
+  const { previewLink } = useStory(({ state: { story: { previewLink } } }) => ({
+    previewLink,
+  }));
+  const [showDialog, setShowDialog] = useState(false);
+  const [isFirstTime, setIsFirstTime] = useState(true);
+  const [user, setUser] = useState({});
+  const [nonce, setNonce] = useState('');
+  const { enablePostLocking } = useFeatures();
+
+  const closeDialog = useCallback(() => {
+    setShowDialog(false);
+    setStoryLockById(storyId);
+  }, [setStoryLockById, storyId]);
+
+  const doGetStoryLock = useCallback(() => {
+    if (enablePostLocking && showLockedDialog) {
+      getStoryLockById(storyId)
+        .then((result) => {
+          if (result.locked && result.user !== userId) {
+            setShowDialog(true);
+            setUser(result['_embedded'].author[0]);
+          } else {
+            setStoryLockById(storyId);
+          }
+          setNonce(result.nonce);
+        })
+        .catch((err) => {
+          trackError('post_lock', err.message);
+        });
+    }
+  }, [
+    getStoryLockById,
+    setStoryLockById,
+    setShowDialog,
+    storyId,
+    userId,
+    enablePostLocking,
+    showLockedDialog,
+  ]);
+
+  // Cache it to make it stable in terms of the below timeout
+  const cachedDoGetStoryLock = useRef(doGetStoryLock);
+  useEffect(() => {
+    cachedDoGetStoryLock.current = doGetStoryLock;
+  }, [doGetStoryLock]);
+
+  useEffect(() => {
+    function releasePostLock() {
+      if (enablePostLocking && showLockedDialog && !showDialog && nonce) {
+        deleteStoryLockById(storyId, nonce);
+      }
+    }
+
+    window.addEventListener('beforeunload', releasePostLock);
+
+    return () => {
+      window.removeEventListener('beforeunload', releasePostLock);
+    };
+  }, [
+    deleteStoryLockById,
+    storyId,
+    enablePostLocking,
+    showLockedDialog,
+    showDialog,
+    nonce,
+  ]);
+
+  useEffect(() => {
+    cachedDoGetStoryLock.current();
+    const timeout = setInterval(() => {
+      cachedDoGetStoryLock.current();
+      setIsFirstTime(false);
+    }, postLockInterval * 1000);
+
+    return () => clearInterval(timeout);
+  }, [postLockInterval]);
+
+  if (isFirstTime) {
+    return (
+      <PostLockFirstDialog
+        open={showDialog}
+        user={user}
+        onClose={closeDialog}
+        previewLink={previewLink}
+        dashboardLink={dashboardLink}
+      />
+    );
+  }
+
+  return (
+    <PostLockDialog
+      open={showDialog}
+      user={user}
+      dashboardLink={dashboardLink}
+    />
+  );
+}
```

Comment:
The `isFirstTime` thing is clever, but not sure if it works reliably.

Also, I would rename the two dialog components to make it more clear what they do.

Perhaps `PostLockDialog` / `PostTakeOverDialog`

Reply:
> The isFirstTime thing is clever, but not sure if it works reliably.

Any idea on a reliably way of doing this?

Reply:
Changed.